Source/Core/Core/State.cpp

```diff
@@ -354,6 +306,71 @@ static std::map<double, int> GetSavedStates()
   return m;
 }
 
+static void CompressBufferToFile(const u8* raw_buffer, u64 size, File::IOFile& f)
+{
+  u64 total_bytes_compressed = 0;
+
+  while (true)
+  {
+    u64 bytes_left_to_compress = size - total_bytes_compressed;
+
+    int bytes_to_compress =
+        static_cast<int>(std::min((u64)LZ4_MAX_INPUT_SIZE, bytes_left_to_compress));
+    int compressed_buffer_size = LZ4_compressBound(bytes_to_compress);
+    auto compressed_buffer = std::make_unique<char[]>(compressed_buffer_size);
+    s32 compressed_len =
+        LZ4_compress_default((char*)raw_buffer + total_bytes_compressed, compressed_buffer.get(),
+                             bytes_to_compress, compressed_buffer_size);
+
+    if (compressed_len == 0)
+      PanicAlertFmtT("Internal LZ4 Error - compression failed");
+
+    // The size of the data to write is 'compressed_len'
+    f.WriteArray(&compressed_len, 1);
+    f.WriteBytes(compressed_buffer.get(), compressed_len);
+
+    total_bytes_compressed += bytes_to_compress;
+    if (total_bytes_compressed == size)
+      break;
+  }
+}
+
+static void CreateExtendedHeader(StateExtendedHeader& extended_header, size_t uncompressed_size)
+{
+  StateExtendedBaseHeader& base_header = extended_header.base_header;
+  base_header.header_version = EXTENDED_HEADER_VERSION;
+  base_header.compression_type =
+      s_use_compression ? CompressionType::LZ4 : CompressionType::Uncompressed;
+  base_header.payload_offset = COMPRESSED_DATA_OFFSET;
+  base_header.uncompressed_size = uncompressed_size;
+
+  // If more fields are added to StateExtendedHeader, set them here.
+}
+
+static void WriteHeadersToFile(size_t uncompressed_size, File::IOFile& f)
+{
+  StateHeader header{};
+  SConfig::GetInstance().GetGameID().copy(header.legacy_header.game_id,
+                                          std::size(header.legacy_header.game_id));
+  header.legacy_header.time = GetSystemTimeAsDouble();
+
+  header.version_header.version_cookie = COOKIE_BASE + STATE_VERSION;
+  header.version_string = Common::GetScmRevStr();
+  header.version_header.version_string_length = (u32)header.version_string.length();
+
+  StateExtendedHeader extended_header{};
+  CreateExtendedHeader(extended_header, uncompressed_size);
+
+  f.WriteArray(&header.legacy_header, 1);
+  f.WriteArray(&header.version_header, 1);
+  f.WriteString(header.version_string);
+
+  f.WriteArray(&extended_header.base_header, 1);
+  // If StateExtendedHeader is ammended to include more than the base, add WriteBytes() calls here.
+  // Do NOT use WriteArray, as this would prevent future revisions from writing char
+  // arrays stored in the extended header.
```

Comment:
I don't understand what you're saying here.

Reply:
So the old code wrote the header to the file via `f.WriteArray(&header, 1)`. This is essentially a memcpy. What I'm trying to say is that, in order to support a future extended header containing a `std::string`, we can't just memcpy the whole thing. Instead we should use `f.WriteBytes()` for every field in the header. This might be an unnecessary comment though.

Reply:
Yes, I think that comment is unnecessary. What function we use for writing the header is something that can be changed in the future anyway.